### PR TITLE
Marshaller Changes

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.jboss.org/browse/ISPN-10525
https://issues.jboss.org/browse/ISPN-10548
https://issues.jboss.org/browse/ISPN-10578
https://issues.jboss.org/browse/ISPN-10505
https://issues.jboss.org/browse/ISPN-10590

~~https://issues.jboss.org/browse/ISPN-10549~~

### CODE REVIEW

**Comment by ryanemerson** [comment]
run performance tests please

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java [hunk 1]
```diff
@@ -16,6 +17,7 @@
  * @author wburns
  * @since 9.0
  */
+@ProtoDoc(ProtoStreamTypeIds.WRAPPED_BYTE_ARRAY)
```
Comment (anistor) [inline commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: This makes me want to add a proper ```@ProtoTypeId``` annotation, but is good ATM :)

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * <p>
+ * Values must be greater than 1000000, as this is the range reserved for internal infinispan use by the protostream
+ * project.
+ * <p>
+ * If message/enum types are no longer required, the variable should be commented instead of deleted.
+ *
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+public interface ProtoStreamTypeIds {
```
Comment (anistor) [inline commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: I'm not a big fan of making commons the melting pot for everything. But what else can we do?

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java (hunk 1 above)
Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: `@ProtoTypeId` would be very nice for the future

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java (hunk 2 above)
Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: Yeah I agree it's a bit clunky, but if we have a Id class per module the chances of us getting overlapping Ids increases.

**Comment by diegolovison** [comment]
run performance tests please

**Comment by ryanemerson** [comment]
run performance tests please

File: commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java (hunk 1 above)
Comment (anistor) [reply commons/src/main/java/org/infinispan/commons/marshall/WrappedByteArray.java]: The future is near https://issues.jboss.org/browse/IPROTO-111
There's also a PR for upgrading protostream.

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package org.infinispan.commons.marshall;
+
+/**
+ * TypeIds used by protostream in place of FQN message/enum names to reduce payload size.
+ * <p>
+ * ONCE SET VALUES IN THIS CLASS MUST NOT BE CHANGED AS IT WILL BREAK BACKWARDS COMPATIBILITY.
+ * <p>
+ * Values must be greater than 1000000, as this is the range reserved for internal infinispan use by the protostream
```
Comment (anistor) [inline commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: 1000000 is a value I came up after an intense decision making process that involved coin tossing. It's now the perfect time to review that and see if we want to reserve a lower range maybe. One that is large enough and that is also efficiently encoded as an int32 with protobuf's zigzag encoding. The range 1000000 to 1005000 would encode to 3 bytes, which is 1 byte lower than a full 32 bit int, so I thought we're fine. But do we want to be even cheaper?

Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: @anistor As cheap as possible ideally. I was curious why we started at `1000000` and not `0`

Comment (anistor) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: Reserving something for internal use should also take into account external usages. that's why 0 is not a good start for the reserved range. Maybe we should think about selecting a smaller (round) value that encodes to 2 bytes, but I would not go lower.

Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: My thoughts were that now protostream is essential to the core of Infinispan, and will only become more important once the GlobalMarshaller uses it, it makes sense to prioritise Infinispan as the primary user in this case.

An additional bytes per user object is not going to have as great an impact on performance. Whereas, one less byte per Infinispan internal object will have a noticeable impact on marshalling size/performance over time and will impact all users, even those using basic types.

Comment (anistor) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: True. Then we could reserve 0..65535 for internal ISPN use and leave the rest for user types. I'll update the javadocs to reflect that. And also create a jira so this remains recorded.

Comment (ryanemerson) [reply commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: Great, thanks

**Comment by ryanemerson** [comment]
run custom performance tests please
ispn_config https://gist.githubusercontent.com/ryanemerson/40e40efb6e73c8c19c8a5d42c3dfe32f/raw/99ded6d69066168d604705dccf9128cc564b63e2/PR-7153.xml
jdg_config https://gist.githubusercontent.com/ryanemerson/df5fd0c76241b9ced6b7d8ea92e1d924/raw/8bc5fe883060fd0a4d674e41fcf606f83e86f5cd/PR-7153-DG.xml
custom_library
entry_factor 8
skip_build

**Comment by ryanemerson** [comment]
This PR is ready for review. My intention is to change the TypeIDs to start at 0 as discussed on this PR, however I am waiting on the upgrade to protostream `alpha11` so that I can make use of the new `@ProtoTypeId` annotation.

File: commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * If message/enum types are no longer required, the variable should be commented instead of deleted.
+ *
+ * @author Ryan Emerson
+ * @since 10.0
+ */
+public interface ProtoStreamTypeIds {
+
+   int MIN_ID = 0;
+
+   // Commons range 1 -> 1000
```
Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java]: We should comment what ranges produce what payload sizes (ie. 1, 2 or 3 bytes). Pretty much everything can go in the 2 byte range(128-16383), except for heavily used things (WrappedByteArray, PutKeyValueCommand etc) should go in 1 byte (0-127) and rarely used things should go into the 3 byte range memcached (16384+).

File: core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java [hunk 5]
```diff
@@ -249,19 +270,43 @@ private Object wrapUserObject(Object o) {
...
+
+   // Protobuf logic included to avoid requiring a dependency on com.google.protobuf.CodedOutputStream
+   private int computeUInt32SizeNoTag(int value) {
+      if ((value & -128) == 0) {
+         return 1;
+      } else if ((value & -16384) == 0) {
+         return 2;
+      } else if ((value & -2097152) == 0) {
+         return 3;
+      } else {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java]: Should this throw an `InvalidArgumentException`? I thought ids can only be 65K or less?

**Comment by wburns** [comment]
Changes seem fine to me. However one thing to note is that we have to be SUPER careful about the performance of JGroupsAddress serialization. This is in every message and can be in some dozens of times when it comes to things like ST messages.

**Comment by wburns** [comment]
Integrated into master, thanks @ryanemerson !

**Comment by anistor** [comment]
Thanks @ryanemerson !

**Pull request merged**